ercc/registry/registry.go

```diff
@@ -0,0 +1,255 @@
+}
+
+// returns a set of credentials registered for a given chaincode id
+// Note: to get the endpoints of FPC endorsing peers do the following:
+// - discover all endorsing peers (and their endpoints) for the FPC chaincode using "normal" lifecycle
+// - query `getEnclaveId` at all the peers discovered
+// - query `queryListEnclaveCredentials` with all received enclave_ids
+// this gives you the endpoints and credentials including enclave_vk, and chaincode_ek
+func (rs *Contract) QueryListEnclaveCredentials(ctx contractapi.TransactionContextInterface, chaincodeId string) ([][]byte, error) {
+	log.Printf("QueryListEnclaveCredentials")
+
+	iter, err := ctx.GetStub().GetStateByPartialCompositeKey("namespaces/credentials", []string{chaincodeId})
+	if iter != nil {
+		defer iter.Close()
+	}
+	if err != nil {
+		return nil, err
+	}
+
+	var allCredentials [][]byte
+	for iter.HasNext() {
+		q, err := iter.Next()
+		if err != nil {
+			return nil, err
+		}
+
+		allCredentials = append(allCredentials, q.Value)
+	}
+
+	return allCredentials, nil
+}
+
+func (rs *Contract) QueryEnclaveCredentials(ctx contractapi.TransactionContextInterface, chaincodeId, enclaveId string) ([]byte, error) {
+	log.Printf("QueryEnclaveCredentials")
+
+	key, err := ctx.GetStub().CreateCompositeKey("namespaces/credentials", []string{chaincodeId, enclaveId})
+	if err != nil {
+		return nil, err
+	}
+
+	credentials, err := ctx.GetStub().GetState(key)
+	if err != nil {
+		return nil, err
+	}
+
+	return credentials, nil
+}
+
+// Optional Post-MVP;
+// returns a list of all provisioned enclaves for a given chaincode id. A provisioned enclave is a registered enclave
+// that has also the chaincode decryption key.
+func (rs *Contract) QueryListProvisionedEnclaves(ctx contractapi.TransactionContextInterface, chaincodeId string) ([]string, error) {
+
+	iter, err := ctx.GetStub().GetStateByPartialCompositeKey("namespaces/credentials", []string{chaincodeId})
+	defer iter.Close()
+	if err != nil {
+		return nil, err
+	}
+
+	var enclaveIds []string
+
+	for iter.HasNext() {
+		q, err := iter.Next()
+		if err != nil {
+			return nil, err
+		}
+
+		_, res, err := ctx.GetStub().SplitCompositeKey(q.Key)
+		if err != nil {
+			return nil, err
+		}
+
+		enclaveId := res[1]
+
+		// next check that for each enclaveID there also exists a CCKeyRegistrationMessage
+		k, err := ctx.GetStub().CreateCompositeKey("namespaces/provisioned", []string{chaincodeId, enclaveId})
+		if err != nil {
+			return nil, err
+		}
+
+		p, err := ctx.GetStub().GetState(k)
+		if err != nil {
+			return nil, err
+		}
+
+		if p != nil {
+			enclaveIds = append(enclaveIds, enclaveId)
+		}
+	}
+
+	return enclaveIds, nil
+}
+
+// returns the chaincode encryption key for a given chaincode id
+func (rs *Contract) QueryChaincodeEncryptionKey(ctx contractapi.TransactionContextInterface, chaincodeId string) ([]byte, error) {
+	//input chaincodeId string
+
+	//return chaincode_ek []byte
+	return nil, nil
```

Comment:
Hmm, some TODO is missing here (or even better, the corresponding implementation: should be straightforward?

Reply:
I added a TODO.

Reply:
Can't we just implement it? should be pretty similar to the other queries?

Reply:
I can be your PR once this here is merged!

Reply:
well, we can do it in a separate PR; just seemed random why we do all the others but not these and add the overhead of a separate PR? But i guess as long as we have the TODO my main concern is addressed